grouper/ctl/base.py

```diff
@@ -1,29 +1,43 @@
 from abc import ABCMeta, abstractmethod
 from typing import TYPE_CHECKING
 
+from grouper.graph import Graph
+from grouper.models.base.session import get_db_engine, Session
+from grouper.repositories.factory import RepositoryFactory
+from grouper.services.factory import ServiceFactory
+from grouper.settings import settings
+from grouper.usecases.factory import UseCaseFactory
+from grouper.util import get_database_url
+
 if TYPE_CHECKING:
     from argparse import _SubParsersAction, Namespace
-    from grouper.models.base.session import Session
+    from typing import Optional
 
 
 class CtlCommand(object):
-    """Implements a subcommand of grouper-ctl."""
+    """Implements a subcommand of grouper-ctl that needs a session and a graph."""
 
     __metaclass__ = ABCMeta
 
+    @staticmethod
     @abstractmethod
-    def add_parser(self, subparsers):
-        # type: (_SubParsersAction) -> None
-        """Add the parser for this subcommand."""
+    def add_parser(subparsers):
+        # type: (_SubParsersAction) -> str
+        """Add the parser for this subcommand and return the subcommand name."""
         pass
 
+    def __init__(self, session):
```

Comment:
It feels like we should either always inject the session and the graph, or go one level higher and only inject the use case factory here. That being said, we can wait until we move over other commands to the new style to see if something emerges organically.

Reply:
I looked at that, but there are two problems. One is that graph construction is really slow (so we would be slowing down every command a **lot** until we fix that). The other is that some commands explicitly want to **not** instantiate a session or the graph before running the command (`sync_db`, for instance).

I think the right fix for this is a factory. I came up with a design on the way into work and will put up a new diff in a moment for you to look at.

Reply:
Take a look at the latest diff and see if that construction looks better to you.